docs/src/main/asciidoc/reaugmentation.adoc

```diff
@@ -0,0 +1,66 @@
+////
+This guide is maintained in the main Quarkus repository
+and pull requests should be submitted there:
+https://github.com/quarkusio/quarkus/tree/main/docs/src/main/asciidoc
+////
+= Reaugment a Quarkus Application
+
+include::./attributes.adoc[]
+
+== What is augmentation?
+
+During augmentation the application is initialized - or let's say pre-initialized.
+Things that used to happen when an EAR file was deployed on a Jakarta-EE server such as parsing static configuration, creating proxy instances, etc. now happen at augmentation time.
+If you add CDI beans after augmentation they won't work (because of the missing proxy classes) and if you change build-time properties (e.g. `quarkus.datasource.db-kind`) after augmentation these changes will be ignored.
+Build-time properties are marked with a lock in the list of link:all-config[All Configuration Options].
+It doesn't matter if you use profiles or any other way to override the properties.
+The build-time properties that were active during augmentation are baked into the build.
+
+> Reaugmentation allows you to recreate the augmentation output for a different build configuration
+
+== When is reaugmentation useful?
+
+Reaugmentation is useful if you ship your software to multiple users which want to be able to change some build-time properties.
+For instance changing the database driver or switching features on or off (e.g. jaeger).
+If there are only two or three possible settings, you can easily provide multiple builds. But if the possible combinations get more you probably prefer shipping a mutable jar (the number of combinations of n switches is 2^n).
+However, you won't be able to use native images with the package type `mutable-jar`.
+Think of the consequences and what other options you have!
+
+It is not a good idea to do reaugmentation at runtime unless you miss the good old times when starting up a server took several minutes and you could enjoy a cup of coffee until it was ready.
+
+== How to reaugment a Quarkus application
+
+In order to run the augmentation steps you need the deployment-jars of the used Quarkus extensions.
+These jars are only present in the mutable-jar distribution. This means that you need to build your application with `quarkus.package.type=mutable-jar`.
+The `mutable-jar` distribution is the same as the `fast-jar` distribution, except for the additional folder `quarkus-app/lib/deployment`
+which contains the deployment-jars and their dependencies (and some class-loader configuration).
+
+TIP: By default, you'll get a warning if a built-time property has been changed at runtime.
+You may set the property `quarkus.configuration.build-time-mismatch-at-runtime=fail` to make sure that you run your application does not startup if there is a mismatch.
+However, as of this writing changing `quarkus.datasource.db-kind` at runtime did neither fail nor produce a warning but was silently ignored.
```

Comment:
@Sanne just in case, is the above behavior expected?

Reply:
No - I wasn't aware and we should fix it.

Reply:
I opened issue https://github.com/quarkusio/quarkus/issues/16786